### PR TITLE
ImportOrderingRule didn't remove blank lines when order was already correct

### PR DESCRIPTION
**Pull request by gabrielittner**
The `testFormatBlankLines` and `testFormatBlankLinesForMajorGroups` tests would have failed without the new `hasTooMuchWhitespace` check inside the rule.

### CODE REVIEW

**Comment by gabrielittner** [comment]
@shyiko did you see this?

**Comment by gabrielittner** [comment]
@shashachu I've rebased this, could you please take a look?

File: ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt [hunk 1]
```diff
@@ -23,8 +24,8 @@ class ImportOrderingRule : Rule("import-ordering") {
             if (children.isNotEmpty()) {
                 val imports = children.filter { it.elementType == IMPORT_DIRECTIVE }
                 val sortedImports = imports.sortedBy { it.text }
-                if (imports != sortedImports) {
-                    emit(node.startOffset, "Imports must be ordered in lexicographic order", true)
+                if (imports != sortedImports || hasTooMuchWhitespace(children)) {
+                    emit(node.startOffset, "Imports must be ordered in lexicographic order in a single group", true)
```
Comment (NightlyNexus) [inline ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: why make the distinction `in a single group`? the rule just takes into account alphabetical order and doesn't care anything about groupings, right?

File: ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt [hunk 2]
```diff
@@ -21,4 +24,94 @@ class ImportOrderingRuleTest {
...
+    private val formattedImports =
+        """
+        import android.app.Activity
+        import android.view.View
+        import android.view.ViewGroup
+        import java.util.List
+        import kotlin.concurrent.Thread
+        """.trimIndent()
+
+    private val errorMessage = "Imports must be ordered in lexicographic order in a single group"
```
Comment (NightlyNexus) [inline ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt]: fwiw, tests are good places for redundancy to keep tight readability and little abstraction. i tend to inline most things. don't know about precedent, though, in this project.

File: ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt [hunk 3]
```diff
@@ -21,4 +24,94 @@ class ImportOrderingRuleTest {
...
+            import java.util.List
+            """.trimIndent()
+        assertThat(ImportOrderingRule().lint(imports)).isEqualTo(
+            listOf(LintError(1, 1, "import-ordering", errorMessage))
+        )
+        assertThat(ImportOrderingRule().format(imports)).isEqualTo(formattedImports)
+    }
+
+    @Test
+    fun testFormatBlankLinesForMajorGroups() {
```
Comment (NightlyNexus) [inline ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt]: redundant with `testFormatBlankLines`?

File: ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt [hunk 4]
```diff
@@ -38,4 +39,8 @@ class ImportOrderingRule : Rule("import-ordering") {
             }
         }
     }
+
+    private fun hasTooMuchWhitespace(nodes: Array<ASTNode>): Boolean {
+        return nodes.any { it is PsiWhiteSpace && (it as PsiWhiteSpace).text != "\n" }
```
[the remark this answers is not in the document]
Comment (gabrielittner) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: I'm getting this error when I remove the cast:
```
ImportOrderingRule.kt:[44,54] Overload resolution ambiguity:
[ERROR] public final val ASTNode.text: String
[ERROR] public final val PsiWhiteSpace.text: String!
```

File: ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt (hunk 1 above)
Comment (gabrielittner) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: I just wanted to be explicit that the rule will not have any grouping. The rule documentation used to say `Separated by a blank line between each major grouping (android, com, junit, net, org, java, javax).`, which it actually never did and shouldn't do. Also with the addition of the `hasTooMuchWhitespace(children)` check it will now actively remove grouping in all cases. 

But I'm ok with removing the addition again.

File: ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt (hunk 2 above)
Comment (gabrielittner) [reply ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt]: I'll inline them.

File: ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt (hunk 3 above)
Comment (gabrielittner) [reply ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt]: It is, I think I did it for explicitness that major groupings will be removed, but it's been some time since I wrote it. I can remove it if you want.

File: ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt (hunk 2 above)
Comment (gabrielittner) [reply ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt]: f39effaa70b144caa1e41a92f7d71727a6999904

File: ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (NightlyNexus) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: oh, interesting. thanks.

File: ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt (hunk 1 above)
Comment (NightlyNexus) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: ah, i didn't have the historical context.
seems like `Imports must be ordered in lexicographic order.` is clear, since there is only one group, but i'm still learning about the project.

File: ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt (hunk 3 above)
Comment (gabrielittner) [reply ktlint-ruleset-standard/src/test/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRuleTest.kt]: Removed it 918bda59f0d730aa927b00569f3d15a4f2ba9384

File: ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt (hunk 1 above)
Comment (gabrielittner) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: Removed it 918bda5

File: ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt [hunk 5]
```diff
@@ -23,7 +24,7 @@ class ImportOrderingRule : Rule("import-ordering") {
             if (children.isNotEmpty()) {
                 val imports = children.filter { it.elementType == IMPORT_DIRECTIVE }
                 val sortedImports = imports.sortedBy { it.text }
-                if (imports != sortedImports) {
+                if (imports != sortedImports || hasTooMuchWhitespace(children)) {
                     emit(node.startOffset, "Imports must be ordered in lexicographic order", true)
```
Comment (shashachu) [inline ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: @gabrielittner Since you're now doing two checks here, I'd at minimum update the error message to say something about white space, or for maximum clarity, have two separate error messages. As it is, a developer would get a failure for a lexographically-ordered import list that happens to have too many white spaces, so that could be confusing.

Comment (shashachu) [inline ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: Could we also move this entire rule under the experimental ruleset? Can also be done as a separate PR.

Comment (gabrielittner) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: I changed the message to `Imports must be ordered in lexicographic order in a single group` originally, but reverted in based on Eric's comments above. I wouldn't mind adding something again.

Comment (gabrielittner) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: I will do that as soon as this one is merged.

Comment (shashachu) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: I don't have strong opinions about the exact language, but I do think we should be specific about what's causing a rule to fail.

If i had
```import a.b.c

import a.b.c.d
import a.b.c.d.e```

and it failed me saying that the imports had to be in lexographic order, I would feel annoyed at minimum, or at worst think the rule was broken.

Comment (shashachu) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: sounds good.

Comment (gabrielittner) [reply ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ImportOrderingRule.kt]: Changed the message to `Imports must be ordered in lexicographic order without any empty lines in-between`

**Comment by shashachu** [comment]
Thanks @gabrielittner!
